llpc/context/llpcCompiler.cpp

```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
   }
 
   if (result != Result::Success) {
-    result = buildPipelineInternal(context, shaderInfo, /*unlinked=*/false, pipelineElf, stageCacheAccesses);
+    // See if we want to use part-pipeline compilation, if there is an FS and at least one pre-rasterization
+    // shader stage.
+    unsigned stageMask = context->getShaderStageMask();
+    bool buildPartPipeline = (cl::EnablePartPipeline && (stageMask & (1 << ShaderStageFragment)) &&
+                              (stageMask & ~(1 << ShaderStageFragment)));
+    if (buildPartPipeline)
+      result = buildGraphicsPipelineWithPartPipelines(context, shaderInfo, pipelineElf, stageCacheAccesses);
+    else
+      result = buildPipelineInternal(context, shaderInfo, PipelineLink::WholePipeline, nullptr, pipelineElf, stageCacheAccesses);
   }
   releaseContext(context);
   return result;
 }
 
+// =====================================================================================================================
+// Build graphics pipeline using part-pipeline compilation
+//
+// @param graphicsContext : Graphics context this graphics pipeline
+// @param shaderInfo : Shader info of this graphics pipeline
+// @param [out] pipelineElf : Output Elf package
+// @param [out] stageCacheAccesses : Stage cache access result. All elements must be initialized in the caller as
+//                                   CacheAccessInfo::CacheNotChecked
+Result Compiler::buildGraphicsPipelineWithPartPipelines(Context *context,
+                                                        ArrayRef<const PipelineShaderInfo *> shaderInfo,
+                                                        ElfPackage *pipelineElf,
+                                                        MutableArrayRef<CacheAccessInfo> stageCacheAccesses) {
+  Result result = Result::Success;
+  unsigned wholeStageMask = context->getShaderStageMask();
+  ElfPackage partPipelineBuffers[2];
+  std::unique_ptr<Pipeline> elfLinkerPipeline;
+  std::unique_ptr<ElfLinker> elfLinker;
+
+  // For each of the two part pipelines (fragment first, then pre-rasterization)...
+  for (unsigned partIdx = 0; partIdx != 2; ++partIdx) {
```

Comment:
Could we change this to iterate over an enum?  Then checks like "partIdx ==0" could read `part == Fragment` or `part == PreRasterization`.  At the vert least we should have named constants for these checks.

Reply:
Done.